### PR TITLE
[FrameworkBundle] Integrate custom property accessors into the framework

### PR DESCRIPTION
**Pull request by lrlopez**
| Q | A |
| --- | --- |
| Branch | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | none |
| License | MIT |
| Doc PR | https://github.com/symfony/symfony-docs/pull/6400 |

This is the last of the two PR related to #18016. It implements changes on `FrameworkBundle` so it integrates with the new features of `PropertyAccess`.

Builds will fail because the new component code has not been merged in this branch in order to ease the review. Should the review be positive, I will rebase this with #22190 so they can work together.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: We prefer the `::class` constant for all new code (same for all class names below)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Most of this code and the one in `registerValidatorMapping` could be factorized in a private method.

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (lrlopez) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I forgot about that one. I'll replace them

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (lrlopez) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Sure!

**Comment by lrlopez** [comment]
@dunglas Done!

**Comment by lrlopez** [comment]
Minor correction to RegisterComponentMapping so it can be merged to 3.4. Failures in the builds seem to be unrelated to the PR.

**Comment by lrlopez** [comment]
@dunglas time for another review? I'm leaving for holidays in a few days and I'd like to close this PR along the way

**Comment by lrlopez** [comment]
@fabpot Fabien, are you still interested in including this feature in Symfony 3.4? Kévin seems to be busy and feature-freeze deadline is almost here. I'd rather not miss it again...

Should that be the case I'll rework the PR so it can be rebased to the 3.4 branch. Thanks!

**Comment by lrlopez** [comment]
I'm working on turning the builds green...

**Comment by lrlopez** [comment]
I'm puzzled with the failed test on PHP 7.x. Can somebody throw some light on how debug it?

**Comment by xabbuh** [comment]
You will need to add the PropertyAccess component to the `composer.json` file of FrameworkBundle with a minimum required version of 3.4 (i.e. add something like `'symfony/property-access': '~3.4|~4.0'` to the `require-dev` section).

**Comment by lrlopez** [comment]
I've tried that, but the failed test is still there. Well, changes to `PropertyAccess` component haven't been merged into the 3.4|4.0 branch yet, so I think is futile: a chicken and egg situation. We'll try to merge a copy of the changes into the repository in order to cheat composer. If it works I'll post the link to Travis-CI test results

**Comment by xabbuh** [comment]
Our Travis config makes sure that the components are taken from the PR for the `deps=low` build. The failure must have been something else then.

**Comment by lrlopez** [comment]
@xabbuh ,
Then I'm confused... Tests pass now when I force the updated `PropertyAccess` component in `composer.json` 😕

File: src/Symfony/Bundle/FrameworkBundle/composer.json
Comment (lrlopez) [inline src/Symfony/Bundle/FrameworkBundle/composer.json]: These changes are only temporary until the PR is accepted or until PR #22190 is merged...

**Comment by nicolas-grekas** [comment]
blocked by pending review on #22190 
@lrlopez this might be moved to 4.1 because we're already late in feature freeze time and the PR is significant
Would it be an issue to merge #22190 in terms of BC? Or does this one adds on top seamlessly, without changing any public/protected interface?

**Comment by lrlopez** [comment]
@nicolas-grekas #22190 does not introduce any BC issues (or at least it should not). There is an additional (and optional) parameter into the constructor of a `PropertyAccessor` class, but if nothing is used the behaviour is 100% equivalent to the previous one.

#22190 was already reviewed by @HeahDude and was only pending on some changes that are already addressed.

This feature will be useful to my students, as they are not proficient in English and find problems with the current singularization rules when creating entities. They'll be very grateful it could be merged into 3.4 which is LTS. I'll commit it as many personal time as I'm able to if necessary to help you on that. Thanks!

**Comment by xabbuh** [comment]
> This feature will be useful to my students, as they are not proficient in English and find problems with the current singularization rules when creating entities.

For that case, wouldn't #23789 be of use for them too? This makes me wonder if we really need both PRs #22190 and #23789. And if we don't need both we should make a careful decision of which PR to merge.

**Comment by lrlopez** [comment]
> For that case, wouldn't #23789 be of use for them too? This makes me wonder if we really need both PRs #22190 and #23789. And if we don't need both we should make a careful decision of which PR to merge.

Ok, I see the struggle here.

The problem with #23789 is that you must create a new naming strategy per project even if you're interested in just overriding just a pair of accessors. It would be a little overkill for my students, but is a viable solution. In fact, my first approach a couple of years ago was similar to David's one (see PR #15766) but Fabien decided that PropertyAccess options was the way to go (#5013). That was the seed of #22190.

Both #23789 and #22190 address some issues in a different way (i.e. singularization) and in any scenario one could be better than the other one, but there is no silver bullet that works best on every scenario.

Each of the approaches have different pros & cons. Let's start with some of them and we can add more as the discussion moves forward:

Custom naming-strategy (#23789)
--------------------------------------
Pros:
- Much simpler.
- Affects all the accessors at once.
- You can extend (or partially replace) the original behaviour by extending a class, which is intuitive.

Cons:
- Doesn't allow to implement virtual properties.
- Changing just a particular accessor implies writing PHP code.
- The custom naming strategy must be global to the project, which could potentially affect other bundles.

Custom property accessors (#22190)
--------------------------------------
Pros:
- Allows defining virtual properties (i.e. calculated ones).
- If using annotations, configuration is next to the affected property or method (is also possible to configure with XML or YAML).
- It does not unintendedly affect any other bundles because it doesn't change the default property access behaviour.

Cons:
- Includes a lot (and I mean a lot) of code changes because it needs to extract metadata from annotations, XML and YAML (in fact most code changes are related to this). For speed reasons, also a cache pool must be used.
- Can't affect all the accessors at the same time, you must define them case-by-case.

Anyway, I don't think both approaches are incompatible but in fact complement each other. Granularity is the key here: sometimes a custom (and global) naming strategy could be the best solution and in others changing a couple of accessors could be just enough.

**Comment by DavidBadura** [comment]
Another use case for NamingStratgy: controlling of camelizing (exists this Word?)

I use the Serializer (ObjectNormalizer) a lot. There is a problem with the interplay of PropertyInfo, PropertyAccess and Serializer (see PR #23656). 

My suggestion was to add a new option "don't camelize", so you can create a special PropertyAccessor for the Serializer component. But one flag for one special usecase is bad. Instead of adding a flag, with #23789 you can change the strategy.

------

But as @lrlopez says, the features complement each other perfectly. I also have use cases where annotation is much easier and better. And would love to see this in the core. But not everything can be solved well with annotation. Actually, everything you want to change Global (for a PropertyAccess instance).

**Comment by lrlopez** [comment]
Until a decision is made, do I need to make any additional changes to the PR? The "Needs work" label is still assigned to the issue

**Comment by chalasr** [comment]
Status updated and milestone changed to 4.1 where PHP7.1 features can be used.

**Comment by nicolas-grekas** [comment]
I'm closing because this staled and something like #22190 should be added first IIUC.
@lrlopez (or anyone else really) PR welcome if you're still interested in seeing this!

**Pull request closed without merging**
